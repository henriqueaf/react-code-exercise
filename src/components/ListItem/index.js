import React, { memo } from 'react';
import MediaLink from '../MediaLink';

const ListItem = ({ member, setSelectedMemberForDetails }) => {
  const {
    id,
    suffix,
    first_name,
    middle_name,
    last_name,
    twitter_account,
    facebook_account,
    youtube_account,
    party,
    district,
    state,
    next_election
  } = member;

  const fullName = () => (
    `${suffix? `${suffix} ` : ''}${first_name}${middle_name? `${middle_name} ` : ' '}${last_name}`
  );

  const districtAndState = () => (
    `${district? `${district}/` : ''}${state}`
  );

  const handleDetailsClick = (member) => (event) => {
    event.preventDefault();
    setSelectedMemberForDetails(member);
  }

  return (
    <tr>
      <td>{id}</td>
      <td>{fullName()}</td>
      <td><MediaLink type="twitter" account_name={twitter_account} /></td>
      <td><MediaLink type="facebook" account_name={facebook_account} /></td>
      <td><MediaLink type="youtube" account_name={youtube_account} /></td>
      <td>{party}</td>
      <td>{districtAndState()}</td>
      <td>{next_election}</td>
      <td><a href="#none" onClick={handleDetailsClick(member)}>Details</a></td>
    </tr>
  );
};

export default memo(ListItem);
