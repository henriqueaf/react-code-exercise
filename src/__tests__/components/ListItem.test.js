import {render, fireEvent, screen} from '@testing-library/react'
import ListItem from '../../components/ListItem';

const member = {
  id: 'A00123',
  suffix: 'Sr.',
  first_name: 'Lamar',
  middle_name: 'Adam',
  last_name: 'Frank',
  twitter_account: 'lamartwt',
  facebook_account: 'lamarfc',
  youtube_account: 'lamaryoutube',
  party: 117,
  district: 'LA',
  state: 'LA',
  next_election: 2022
}

const mockSetSelectedMemberForDetails = jest.fn(() => {});

const defaultProps = {
  member,
  setSelectedMemberForDetails: mockSetSelectedMemberForDetails
};

const mount = (props) => {
  const mergedProps = {
    ...defaultProps,
    ...props,
  };

  const tbody = document.createElement('tbody');
  return render(<ListItem {...mergedProps} />, {container: document.body.appendChild(tbody)});
};

describe('ListItem', () => {
  it('Matches snapshot', () => {
    const {container} = mount();
    expect(container).toMatchSnapshot();
  });

  it('Renders a tr with some td inside', () => {
    mount();
    const memberTableDatas = screen.getAllByRole('cell');
    expect(memberTableDatas.length).toBe(9);
  });

  it('Renders member id', () => {
    mount();
    const memberIdTableData = screen.getAllByRole('cell')[0];
    expect(memberIdTableData.textContent).toBe(member.id);
  });

  it('Trigger setSelectedMemberForDetails when details link is pressed', () => {
    mount();
    const link = screen.getByText('Details');

    fireEvent.click(link, {preventDefault: () => {}});
    expect(mockSetSelectedMemberForDetails).toBeCalledTimes(1);
  });
})
