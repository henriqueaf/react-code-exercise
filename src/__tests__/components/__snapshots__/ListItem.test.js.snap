// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem Matches snapshot 1`] = `
<tbody>
  <tr>
    <td>
      A00123
    </td>
    <td>
      Sr. LamarAdam Frank
    </td>
    <td>
      <a
        href="https://twitter.com/lamartwt"
        rel="noreferrer"
        target="_blank"
      >
        lamartwt
      </a>
    </td>
    <td>
      <a
        href="https://www.facebook.com/lamarfc"
        rel="noreferrer"
        target="_blank"
      >
        lamarfc
      </a>
    </td>
    <td>
      <a
        href="https://www.youtube.com/user/lamaryoutube"
        rel="noreferrer"
        target="_blank"
      >
        lamaryoutube
      </a>
    </td>
    <td>
      117
    </td>
    <td>
      LA/LA
    </td>
    <td>
      2022
    </td>
    <td>
      <a
        href="#none"
      >
        Details
      </a>
    </td>
  </tr>
</tbody>
`;
